Add dragging behaviour properties to sliders

ibdafna opened the pull request:
This PR adds functionality requested in #2832 . The drag behaviour API has been exposed on the python side and now allows four selections:

1. Tap (current default behaviour)
2. Snap (new option)
3. Drag-tap (new default behaviour for all sliders - no change in behaviour for non-range sliders)
4. Drag-snap

There are no breaking API changes, but a new `behaviour` attribute has been added to the python side.

![devw_drag_behaviour1](https://user-images.githubusercontent.com/24281433/77391612-707bd480-6d56-11ea-92e3-b7b28676eb22.gif)

jasongrout commented:
> This PR adds functionality requested in #2683 .

#2832, right?

I was thinking it might be better to start out with just changing the default behavior to `'drag-tap'`, which addresses the user request, and not expose changing it to the users just yet. The tradeoff here is that increasing API surface means we ought to support it in the future (even if we change slider), so it becomes a maintenance burden. On the other hand, it might be useful enough to users to have a tap/snap/drag-snap behavior available. Are there usecases people actually see that do *not* want drag-tap behavior?

jasongrout commented:
If exposing this behavior to the users is the way we want to go, then I think you've done a great job here (just reading the code, I haven't run it yet)

ibdafna commented:
> > This PR adds functionality requested in #2683 .
> 
> #2832, right?

Indeed. Updated the PR note above with the correct one, thanks for spotting!

> I was thinking it might be better to start out with just changing the default behavior to `'drag-tap'`, which addresses the user request, and not expose changing it to the users just yet. The tradeoff here is that increasing API surface means we ought to support it in the future (even if we change slider), so it becomes a maintenance burden. On the other hand, it might be useful enough to users to have a tap/snap/drag-snap behavior available. Are there usecases people actually see that do _not_ want drag-tap behavior?

Potentially yes. Off the top of my head, let's say I'm using the range slider to control the spot range of a vanilla option payoff at expiry. I would probably not want to drag both handles of the slider at the same time given I am only looking to focus on the strike area. But then again, it might be convenient to drag the full range if I got the original parameters wrong and need to shift my window left or right.

> Thoughts? I know I've become very conservative about extending the core ipywidgets api because it aims to be very stable over a long term, so I'm wary of adding maintenance burdens.

I'm definitely with you on robustness. I added the snap behaviour because I often expected the slider handles to behave that way and thought other users might also appreciate that option. If you think we are incurring potential technical debt down the line then I'm happy to remove the changes!

vidartf commented:
> If you think we are incurring potential technical debt down the line then I'm happy to remove the changes!

We could also consider to "expose" it as an underscore param `_behavior` ("use at own risk"). 🤷‍♂

agregory96 commented:
> > This PR adds functionality requested in #2683 .
> 
> #2832, right?
> 
> I was thinking it might be better to start out with just changing the default behavior to `'drag-tap'`, which addresses the user request, and not expose changing it to the users just yet. The tradeoff here is that increasing API surface means we ought to support it in the future (even if we change slider), so it becomes a maintenance burden. On the other hand, it might be useful enough to users to have a tap/snap/drag-snap behavior available. Are there usecases people actually see that do _not_ want drag-tap behavior?
> 
> Thoughts? I know I've become very conservative about extending the core ipywidgets api because it aims to be very stable over a long term, so I'm wary of adding maintenance burdens.

This was precisely what I was looking for, thanks much to everyone for the quick response, review, and implementation!

ibdafna commented:
@jasongrout How should we proceed?

jtpio commented:
For folks wanting to try this PR on Binder:

https://mybinder.org/v2/gh/ibdafna/ipywidgets/drag_behaviour?urlpath=lab/tree/docs/source/examples/

jtpio commented on ipywidgets/widgets/widget_float.py:
```diff
@@ -159,6 +159,8 @@ class FloatSlider(_BoundedFloat):
         default is 'horizontal', orientation of the slider
     readout : {True, False}
         default is True, display the current value of the slider next to it
+    behaviour : str
```
Commenting here as this came up during the weekly meeting today, with @ianhi spotting this attribute was not in American English:

```suggestion
    behavior : str
```

jtpio replied:
Ah that was probably because of `noUISlider` using `behaviour` in the options:

https://refreshless.com/nouislider/slider-options/#section-limit

![image](https://user-images.githubusercontent.com/591645/133807331-4f0be361-8526-4097-86d0-227b5f705c7c.png)

jtpio commented:
FYI @ibdafna I rebased and added a commit to rename `behaviour` to `behavior` (American English)

ibdafna commented:
@jtpio thank you 😄
